Library/OcCpuLib/OcCpuLib.c

```diff
@@ -273,33 +273,81 @@ ScanThreadCount (
 
 STATIC
 VOID
-ScanIntelFSBFrequency (
-  IN  OC_CPU_INFO        *CpuInfo
+SetMaxBusRatioAndMaxBusRatioDiv (
+  IN   OC_CPU_INFO        *CpuInfo  OPTIONAL,
+  OUT  UINT8              *MaxBusRatio,
+  OUT  UINT8              *MaxBusRatioDiv
   )
 {
   MSR_IA32_PERF_STATUS_REGISTER       PerfStatus;
   MSR_NEHALEM_PLATFORM_INFO_REGISTER  PlatformInfo;
-  UINT8                               MaxBusRatio;
-  UINT8                               MaxBusRatioDiv;
+  CPUID_VERSION_INFO_EAX              Eax;
+  UINT8                               CpuModel;
 
-  //
-  // TODO: this may not be accurate on some older processors.
-  //
-  if (CpuInfo->CpuGeneration >= OcCpuGenerationNehalem) {
-    PlatformInfo.Uint64 = AsmReadMsr64 (MSR_NEHALEM_PLATFORM_INFO);
-    MaxBusRatio = (UINT8) PlatformInfo.Bits.MaximumNonTurboRatio;
-    MaxBusRatioDiv = 0;
+  if (MaxBusRatio == NULL || MaxBusRatioDiv == NULL) {
+    return;
+  }
+
+  if (CpuInfo != NULL) {
+    //
+    // TODO: this may not be accurate on some older processors.
+    //
+    if (CpuInfo->CpuGeneration >= OcCpuGenerationNehalem) {
+      PlatformInfo.Uint64 = AsmReadMsr64 (MSR_NEHALEM_PLATFORM_INFO);
+      *MaxBusRatio = (UINT8) PlatformInfo.Bits.MaximumNonTurboRatio;
+      *MaxBusRatioDiv = 0;
+    } else {
+      PerfStatus.Uint64 = AsmReadMsr64 (MSR_IA32_PERF_STATUS);
+      *MaxBusRatio = (UINT8) (RShiftU64 (PerfStatus.Uint64, 8) & 0x1FU);
+      //
+      // Undocumented values:
+      // Non-integer bus ratio for the max-multi.
+      // Non-integer bus ratio for the current-multi.
+      //
+      *MaxBusRatioDiv = (UINT8) (RShiftU64 (PerfStatus.Uint64, 46) & BIT0);
+    }
   } else {
-    PerfStatus.Uint64 = AsmReadMsr64 (MSR_IA32_PERF_STATUS);
-    MaxBusRatio = (UINT8) (RShiftU64 (PerfStatus.Uint64, 8) & 0x1FU);
     //
-    // Undocumented values:
-    // Non-integer bus ratio for the max-multi.
-    // Non-integer bus ratio for the current-multi.
+    // Assuming Intel machines used on Apple hardware.
     //
-    MaxBusRatioDiv = (UINT8) (RShiftU64 (PerfStatus.Uint64, 46) & BIT0);
+    AsmCpuid (
+      CPUID_VERSION_INFO,
+      &Eax.Uint32,
+      NULL,
+      NULL,
+      NULL
+      );
+    CpuModel = (UINT8) Eax.Bits.Model | (UINT8) (Eax.Bits.ExtendedModelId << 4U);
+    //
+    // This part should be synced with the code above.
+    //
+    if (CpuModel >= CPU_MODEL_NEHALEM) {
+      PlatformInfo.Uint64 = AsmReadMsr64 (MSR_NEHALEM_PLATFORM_INFO);
+      *MaxBusRatio = (UINT8) PlatformInfo.Bits.MaximumNonTurboRatio;
+      *MaxBusRatioDiv = 0;
+    } else {
+      PerfStatus.Uint64 = AsmReadMsr64 (MSR_IA32_PERF_STATUS);
+      *MaxBusRatio = (UINT8) (RShiftU64 (PerfStatus.Uint64, 8) & 0x1FU);
+      *MaxBusRatioDiv = (UINT8) (RShiftU64 (PerfStatus.Uint64, 46) & BIT0);
+    }
+  }
```

Comment:
Add a check, if `*MaxBusRatio` is 0, assign it 1, and then drop the `MaxBusRatio > 0` check. Sounds like a safe failsafe in my opinion.